### ISSUE TITLE
bootctl: RENAME_NOREPLACE is not supported on old kernels

### ISSUE THREAD

**Issue by tobiasBora** [systemd/systemd#10063]
**Expected behaviour you didn't see**

When running `bootctl --path=/boot install` on a distribution with a quite old kernel (like Ubuntu 16.04), it fails with an error `Failed to move "/boot/loader/loader.conf" into place: Operation not permitted`. As already reported [here](https://github.com/NixOS/nixpkgs/issues/46098) (it also includes a `strace` and full list of command to find this problem),  it is because of the flag `RENAME_NOREPLACE` that is not supported on old kernels, and because FAT32 does not support links, it fails. 

Because bootctl is quite an important software that may be use on lot's of different kind of OS, I think it would be interesting to solve this issue, even if it concerns only old kernels, especially because as far as I understand the problem, it seems pretty easy to solve.

**Unexpected behaviour you saw**

I would expect this command not to fail.

**Steps to reproduce the problem**

Load an Ubuntu 16.04, install nixos as explained in the manual with an UEFI (FAT32) boot partition. It will automatically load bootctl, and fail. For full log, [see](https://github.com/NixOS/nixpkgs/issues/46098).

**Comment by kirbyfan64** [comment]
FWIW 16.04 has kernel 4.4, which is technically supposed to be supported by systemd.

**Comment by poettering** [comment]
ah, that sucks, and i have no idea how to properly emulate that...

**Comment by tobiasBora** [comment]
@poettering I tried with qemu and I got the same error. Steps: first get [xubuntu-16.04.1-desktop-amd64.iso ](http://cdimage.ubuntu.com/xubuntu/releases/16.04/release/xubuntu-16.04.1-desktop-amd64.iso), as well as the two files compressed here to simulate UEFI:
[OVMF.tar.gz](https://github.com/systemd/systemd/files/2379840/OVMF.tar.gz)

Then, put all these files in the same folder and run:

    qemu-system-x86_64 -m 3G -enable-kvm -vga qxl -drive if=pflash,format=raw,readonly,file=OVMF_CODE.fd -drive if=pflash,format=raw,file=OVMF_VARS.fd -drive if=virtio,file=debug_bootctl.qcow2 -cdrom xubuntu-16.04.1-desktop-amd64.iso 

boot in live mode, and using gparted, create a new GPT table partition on the drive vda, one FAT32 of 500M, one ext4, and set the flag esp/boot to the partition. Then run the following commands:

```
sudo mount /dev/vda2 /mnt/
sudo mkdir /mnt/boot/
sudo mount /dev/vda1 /mnt/boot/
sudo apt install curl
bash <(curl https://nixos.org/nix/install)
. $HOME/.nix-profile/etc/profile.d/nix.sh
nix-env -iE "_: with import <nixpkgs/nixos> { configuration = {}; }; with config.system.build; [ nixos-generate-config nixos-install nixos-enter manual.manpages ]"
sudo `which nixos-generate-config` --root /mnt
sudo groupadd -g 30000 nixbld
sudo useradd -u 30000 -g nixbld -G nixbld nixbld
sudo PATH="$PATH" NIX_PATH="$NIX_PATH" `which nixos-install` --root /mnt
```
You should get an error now.

**Comment by lucaswerkmeister** [comment]
Instead of the `link`+`unlink` workaround, would it work to `open` the file, `unlink` it, and then `link` `/proc/self/fd/NUM` into place instead? In theory, this doesn’t require file system support for more than one link to a file at any point, though I’m not sure if the Linux FAT implementation actually supports it.

(This trick is based on the documentation for `O_TMPFILE` in the `open(2)` manpage.)

**Comment by poettering** [comment]
link+unlink doesn't work btw, as that this is FAT we are talking of, which doesn't do hardlinks, and hence won't allow linking a file while it already has a name in the file system...

@lucaswerkmeister I don't think this works. THe fact that you can link /proc/self/fd/$FD into a file system is a special feature of O_TMPFILE iirc, and not supported on other fds. But maybe I am wrong on that, we should check...

**Comment by lucaswerkmeister** [comment]
Looks like you’re right, the `linkat` syscall results in `EPERM` ([gist](https://gist.github.com/lucaswerkmeister/8eced991bccb04b2aa7f0af32a3c40d5)) :(

**Comment by poettering** [comment]
Fix waiting in #10251
